quixstreams/app.py

```diff
@@ -964,3 +968,69 @@ def _on_sigint(self, *_):
     def _on_sigterm(self, *_):
         logger.debug(f"Received SIGTERM, stopping the processing loop")
         self.stop()
+
+
+# TODO: python 3.10
+# @dataclasses.dataclass(frozen=True, slots=True)
+@dataclasses.dataclass(frozen=True)
+class ApplicationConfig:
+    """
+    Immutable object holding the application configuration
+
+    For details see :class:`quixstreams.Application`
+    """
+
+    broker_address: ConnectionConfig
+    consumer_group: str
+    auto_offset_reset: AutoOffsetReset = "latest"
+    commit_interval: float = 5.0
+    commit_every: int = 0
+    producer_extra_config: dict = dataclasses.field(default_factory=dict)
+    consumer_extra_config: dict = dataclasses.field(default_factory=dict)
+    processing_guarantee: ProcessingGuarantee = "at-least-once"
+    consumer_poll_timeout: float = 1.0
+    producer_poll_timeout: float = 0.0
+    auto_create_topics: bool = True
+    on_message_processed: Optional[MessageProcessedCallback] = None
+    request_timeout: float = 30
+    topic_create_timeout: float = 60
+    state_dir: str = "state"
+    rocksdb_options: Optional[RocksDBOptionsType] = None
+    use_changelog_topics: bool = True
+
+    def validate(self) -> None:
+        """
+        Validate the configuration
+        """
+        if self.processing_guarantee not in get_args(ProcessingGuarantee):
+            raise ValueError(
+                f'Must provide a valid "processing_guarantee"; expected one of: '
+                f'{get_args(ProcessingGuarantee)}, got "{self.processing_guarantee}"'
+            )
+
+        if self.auto_offset_reset not in get_args(AutoOffsetReset):
+            raise ValueError(
+                f'Must provide a valid "auto_offset_reset"; expected one of: '
+                f'{get_args(AutoOffsetReset)}, got "{self.auto_offset_reset}"'
+            )
+
+    def copy(self, **kwargs) -> "ApplicationConfig":
```

Comment:
`-> "ApplicationConfig"` can be replaced with `-> Self` from the `typing_extensions` module.